indico/modules/users/forms.py

```diff
@@ -28,6 +28,12 @@
 
 class UserPreferencesForm(IndicoForm):
     lang = SelectField(_('Language'))
+
+    force_language = BooleanField(
+        _('Use my language'),
+        widget=SwitchWidget(),
+        description=_("Always use my preferred language instead of an event's supported language."))
```

Comment:
I'd also include this sentence: "This may result in a mix of two languages in some areas of an event."

Reply:
Hi @ThiefMaster,
I think, If the `force_language` is enabled it will not lead to a mix of the languages, as the mix of the languages will happen when it's disabled and an event does not support the user's language and will use the event's default language. correct?

```
User setting enabled: Always use the user’s language (ignore supported languages); otherwise use the user’s language if it’s supported, and fall back to the default language if it isn’t.
````

Reply:
If "use my language" is disabled, the user will get the event's language - so e.g. registration form field names / conference menu item titles will match the Indico language used for (translatable) text coming from indico.

But if "user my language" is enabled, the user basically says "I always want text from indico to be in my language". so if I sign up for a French event while Indico is set to English, I will get emails that have English texts from indico, but french texts that were provided by the event organizer...

Reply:
The language lookup logic should be this (first match is used):

- User has `force_language` enabled? Use the user's language.
- User's language is supported by the event? Use the user's language
- User's language is not supported by the event? Use the event's default language
- There's no user (so no user language information available either)? Use the event's default language

Reply:
I believe my following code satisfies all the 4 points,

```py
    def force_event_locale(self, user=None):
        locale = self.default_language
        if user:
            locale = user.settings.get('lang')
            if not user.settings.get('force_language') and locale not in self.supported_languages:
                locale = self.default_language

        with force_locale(locale):
            yield
```